File: sdk-js/src/gatewayHmacUserContext.test.ts
import {
  buildGatewayUserContextString,
  buildGatewayUserContextStringV2,
  buildGatewayUserContextStringV3,
  grantsAccess,
} from './verifier';

describe('buildGatewayUserContextStringV3', () => {
  it('all fields present golden string', () => {
    expect(
      buildGatewayUserContextStringV3(
        'sub-ext-id',
        'prod-uuid-1',
        ['roleA', 'roleB'],
        'ACTIVE',
        'SUBSCRIPTION',
        'PER_REQUEST',
        'request'
      )
    ).toBe('3sub-ext-idprod-uuid-1roleA,roleBACTIVESUBSCRIPTIONPER_REQUESTrequest');
  });

  it('empty roles golden string', () => {
    expect(buildGatewayUserContextStringV3('user-1', 'prod-1', [], 'TRIAL', null, null, null)).toBe(
      '3user-1prod-1TRIAL'
    );
  });

  it('billing fields absent golden string', () => {
    expect(buildGatewayUserContextStringV3('owner-id', '', [], 'ACTIVE', null, null, null)).toBe(
      '3owner-idACTIVE'
    );
  });

  it('non-access status golden string', () => {
    expect(
      buildGatewayUserContextStringV3(
        'user-x',
        'prod-x',
        ['r1'],
        'EXPIRED',
        'PREPAID',
        'PER_REQUEST',
        'request'
      )
    ).toBe('3user-xprod-xr1EXPIREDPREPAIDPER_REQUESTrequest');
  });
});

describe('buildGatewayUserContextStringV2 (compat)', () => {
  it('owner no roles', () => {
    expect(buildGatewayUserContextStringV2('user-1', 'owner', [], true, null, null, null)).toBe(
      '2user-1ownertrue'
    );
  });
});

describe('buildGatewayUserContextString v1 (compat)', () => {
  it('with quota', () => {
    expect(buildGatewayUserContextString('a', 'b', ['x'], 5, false, 'S', 'M', 'U')).toBe('abx5falseSMU');
  });
});

describe('grantsAccess', () => {
  it('returns true for invoke-eligible statuses', () => {
    expect(grantsAccess('ACTIVE')).toBe(true);
    expect(grantsAccess('trial')).toBe(true);
    expect(grantsAccess('CANCELLING')).toBe(true);
    expect(grantsAccess('CANCELLING_PENDING')).toBe(true);
  });

  it('returns false for other or missing status', () => {
    expect(grantsAccess('EXPIRED')).toBe(false);
    expect(grantsAccess(null)).toBe(false);
    expect(grantsAccess('')).toBe(false);
  });
});

File: sdk-java/src/test/java/com/tollara/common/util/GatewayHmacUserContextTest.java
package com.tollara.common.util;

import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class GatewayHmacUserContextTest {

    @Test
    void buildV3_allFieldsPresent_goldenString() {
        String ctx = GatewayHmacUserContext.buildV3(
                "sub-ext-id",
                "prod-uuid-1",
                List.of("roleA", "roleB"),
                "ACTIVE",
                "SUBSCRIPTION",
                "PER_REQUEST",
                "request");
        assertEquals("3sub-ext-idprod-uuid-1roleA,roleBACTIVESUBSCRIPTIONPER_REQUESTrequest", ctx);
    }

    @Test
    void buildV3_emptyRoles_goldenString() {
        String ctx = GatewayHmacUserContext.buildV3(
                "user-1",
                "prod-1",
                List.of(),
                "TRIAL",
                null,
                null,
                null);
        assertEquals("3user-1prod-1TRIAL", ctx);
    }

    @Test
    void buildV3_billingFieldsAbsent_goldenString() {
        String ctx = GatewayHmacUserContext.buildV3(
                "owner-id",
                "",
                null,
                "ACTIVE",
                null,
                null,
                null);
        assertEquals("3owner-idACTIVE", ctx);
    }

    @Test
    void buildV3_nonAccessStatus_goldenString() {
        String ctx = GatewayHmacUserContext.buildV3(
                "user-x",
                "prod-x",
                List.of("r1"),
                "EXPIRED",
                "PREPAID",
                "PER_REQUEST",
                "request");
        assertEquals("3user-xprod-xr1EXPIREDPREPAIDPER_REQUESTrequest", ctx);
    }

    @Test
    void build_v2_ownerNoRoles_goldenString() {
        String ctx = GatewayHmacUserContext.buildV2(
                "user-1",
                "owner",
                List.of(),
                true,
                null,
                null,
                null);
        assertEquals("2user-1ownertrue", ctx);
    }

    @Test
    void build_v2_subscriberWithRolesAndBillingFields_goldenString() {
        String ctx = GatewayHmacUserContext.buildV2(
                "sub-ext-id",
                "pro",
                List.of("roleA", "roleB"),
                false,
                "SUBSCRIPTION",
                "PER_REQUEST",
                "request");
        assertEquals("2sub-ext-idproroleA,roleBfalseSUBSCRIPTIONPER_REQUESTrequest", ctx);
    }

    @Test
    void build_legacy_v1_withQuota_goldenString() {
        String ctx = GatewayHmacUserContext.build(
                "a",
                "b",
                List.of("x"),
                new BigDecimal("5"),
                false,
                "S",
                "M",
                "U");
        assertEquals("abx5falseSMU", ctx);
    }
}
